### ISSUE TITLE
getting rid of raw pointers

### ISSUE THREAD

**Issue by jeanluct** [jeanluct/ttauto#5]
I tried, on the branch `try-shared_ptr-multigon`, to convert the remaining pointers to smart pointers.  This has proved difficult.  There's one glaring problem at the outset: mutligons point to edges, but edges point back to multigon.  This is a circular loop, which is very bad for `shared_ptr`.  So the first thing to try is the following:
* `traintrack` has `unique_ptr` to each `multigon`; (this is the current setup)
* `multigon` has a `shared_ptr` to each edge it owns.  This means each edge has two owners, so the reference count is always 2; (this is the current setup)
* each `edge` points to two multigons.  At the outset this was done with raw pointers.  I tried switching to `weak_ptr` but I'm now sure I completely understand how this change things.  You still have to `lock` the pointer (i.e., convert to shared) whenever you want to use it.

In any case the code is not working in that branch (segfaults).  The trickiest function was `swap`ing multigons: it turns out this function depended heavily on getting a reference to multigons rather than a pointer.

Is the solution to get rid of pointers back to multigons completely?  I already have a label set up, so maybe an edge could point back to a multigon through its label.  But maybe this would conflict with the "marked" option.

**Comment by jeanluct** [comment]
Updated execution plan for issue #5 (restart from master)

The previous `try-shared_ptr-multigon` attempt introduced segfaults and should be treated as an experiment branch, not a merge target.

Master currently works, so the plan is to restart from `master` and make small, testable changes.

## Why restart
- The old branch changed ownership and mutation logic too broadly at once.
- `multigon <-> edge` references made lifetime rules unclear during `swap`.
- The branch includes debug scaffolding and known broken behavior.

## Target ownership model
- `traintrack` is the only owner of graph objects.
- Internal links (`edge -> multigon`, `multigon -> edge`) are non-owning.
- Avoid `shared_ptr` cycles in internal graph links.
- If/when IDs are introduced, they should be stable across `swap` operations.

## Incremental execution plan
1. Baseline and reproducer
   - Start from `master`.
   - Add/identify a focused regression test for `swap` and edge consistency.
   - Keep behavior identical before ownership refactors.

2. Document ownership contracts in code
   - Add clear comments in `traintrack`, `multigon`, and `edge` headers:
     - owning vs non-owning references
     - lifetime expectations
   - No behavior changes in this step.

3. Stabilize mutation invariants
   - Add consistency checks around `swap` and edge relinking.
   - Validate:
     - endpoint multigons are valid
     - prong/index metadata remains consistent
     - no detached/internally inconsistent edges after swap

4. Refactor one pointer path at a time
   - Choose one direction first (prefer `edge -> multigon` back-links).
   - Keep interfaces mostly unchanged while replacing ownership semantics.
   - Run tests after each small commit.

5. Optional ID-based back-links (only if needed)
   - If pointer-based non-owning links remain fragile, move back-links to stable labels/indices with lookup through `traintrack`.
   - Do this only after pointer-based invariants are passing.

6. Final hardening
   - Run full `scons` build and all test binaries.
   - Add one stress test with repeated swaps/folds.
   - Remove temporary debug prints and dead experimental code.

## Merge policy for this work
- Only cherry-pick clean, minimal commits into the working branch.
- Do not merge the old experimental pointer-conversion branch wholesale.
- Require: no segfault reproducer, tests pass, invariants pass.

**Comment by jeanluct** [comment]
Issue #5 closure update: restart from master completed and stabilized

I did **not** merge the old experimental shared_ptr-conversion branch. Instead, this was restarted from `master` and implemented as a sequence of small, behavior-preserving commits.

## What was done

### 1) Ownership contract clarified (no broad smart-pointer redesign)
- Documented intended ownership in core headers:
  - `traintrack` is the sole owner of multigons.
  - multigon/edge cross-links are non-owning.
- Kept the existing ownership architecture rather than pushing further into shared ownership for multigons.

### 2) Mutation/relink invariants strengthened
- Added local postcondition checks around relink/swap paths in `multigon` logic.
- Added helper-based consistency checks to keep failure points explicit and deterministic.

### 3) Back-link handling refactored for maintainability
- Incrementally replaced direct edge metadata manipulation in `multigon` call sites with small edge helpers (ending access, renumber, relink).
- Reduced duplication and narrowed low-level back-link writes to edge-centric helpers.

### 4) Test hardening
- Extended `tests/test_traintrack` with a bounded fold->check->normalise->check mutation stress loop.
- Added `TESTING.md` with recommended quick/long-run validation commands.

### 5) Build/runtime hardening
- Updated SCons baseline to C++17.
- Updated SCons linking to use the CSparse static archive directly so examples/tests do not depend on runtime discovery of `libcsparse.so`.

## Validation status
- Full `scons` build passes.
- Core test binaries pass:
  - `tests/test_permplus1`
  - `tests/test_traintrack`
  - `tests/test_folding_path`
  - `tests/test_badwords`
- Example smoke runs pass (`ttauto_min_example`, `ttauto_torus`, default-input run of `examples/ttauto`).

## Conclusion
Given that `master` had no segfault and the segfault was introduced by broad experimental ownership changes, this issue is considered addressed by stabilization + invariants + targeted refactors rather than a full smart-pointer rewrite.

## Possible next phase (optional, later)
If desired in future, a separate follow-up issue can evaluate deeper redesign options (e.g., ID-based back-links, sanitizer-focused CI, or stricter static analysis), but this should be done as narrow, test-first increments from stable `master`.

**Comment by jeanluct** [comment]
Closing as completed via the restart-from-master stabilization series now merged on local master. See the previous comment for full details and optional future directions.

**Comment by jeanluct** [comment]
The two relevant branches are

`iss005-try-shared_ptr-multigon`                 # abandoned
`iss005-try-shared_ptr-multigon_restart`   # merged in master
